src/commands/selection.rs

```diff
@@ -100,6 +123,75 @@ fn copy_to_clipboard(app: &mut Application) -> Result {
     Ok(())
 }
 
+/// Get the selected range from an application in a selection mode. *Requires*
+/// that the application has a buffer and is in mode Select, SelectLine, or
+/// Search.
+fn range_from(app: &mut Application) -> Range {
+    let buffer = app.workspace.current_buffer();
+    let buffer = buffer.unwrap();
+
+    match app.mode {
+        Mode::Select(ref select_mode) => {
+            Range::new(*buffer.cursor.clone(), select_mode.anchor)
+        }
+        Mode::SelectLine(ref select_line_mode) => {
+            select_line_mode.to_range(&*buffer.cursor)
+        }
+        Mode::Search(ref mode) => {
+            mode.results
+            .as_ref()
+            .and_then(|r| r.selection())
+            .ok_or("Cannot get selection outside of select mode.")
+            .unwrap()
+            .clone()
+        }
+        _ => panic!("Cannot get selection outside of select mode."),
+    }
+}
+
+/// Wrap a string at a given maximum length (generally 80 characters). If the
+/// line begins with a comment (matches potential_prefix), the text is wrapped
+/// around it.
+fn justify_string(text: &String, max_len: usize, potential_prefix: Regex) -> String {
+    let mut justified = String::new();
+    for paragraph in text.split("\n\n") {
+        let mut paragraph = paragraph.split_whitespace().peekable();
+        let prefix;
+        let max_len_with_prefix;
+        if paragraph.peek().is_some()
+           && potential_prefix.is_match(paragraph.peek().unwrap())
+        {
+            prefix = paragraph.next().unwrap().to_owned() + " ";
+            max_len_with_prefix = max_len - prefix.len();
+            justified += &prefix;
+        } else {
+            prefix = String::new();
+            max_len_with_prefix = max_len;
+        }
+
+        let mut len = 0;
+
+        for word in paragraph {
+            if word == prefix {
+                continue;
+            }
+
+            len += word.len() + 1;
+            if len > max_len_with_prefix {
+                len = word.len();
+                justified.push('\n');
+                justified += &prefix;
+            }
+            justified += word;
+            justified.push(' ');
+        }
+
+        justified += "\n\n"; // add the paragraph delim
```

Comment:
Why add some additional newlines here? When justifying a paragraph I wouldn't expect it to suddenly add superfluous lines after the paragraph.

Reply:
The whitespace collapsing removes interprets paragraph breaks (\n\n) the same as line breaks and word breaks, so paragraph breaks would be removed without it. However, I can see how that would be very sub-optimal on single paragraphs. I suppose the solution to that is to only add the paragraph breaks if the current paragraph isn't the final. I'll ammend this so that it only runs if the current paragraph is not the last, and add relevant tests.

Thank you for the thorough review, and sorry my code is such a mess!

Reply:
Ah, I understand! Yeah, that makes sense!
Don't worry, that's what review is for. Thanks for doing the work, it's something very useful to have! 👍